### ISSUE TITLE
`TrustRegion` solver has issues when using low-precision (`f32`) type as `Problem::Scalar`

### ISSUE THREAD

**Issue by twitchyliquid64** [datamole-ai/gomez#22]
This solver works:

```rust
struct SolveForOrigin;

impl Problem for SolveForOrigin {
    type Scalar = f64;
    type Dim = na::U2;

    fn dim(&self) -> Self::Dim {
        na::U2::name()
    }
}

impl System for SolveForOrigin {
    fn eval<Sx, Sfx>(
        &self,
        x: &na::Vector<Self::Scalar, Self::Dim, Sx>,
        fx: &mut na::Vector<Self::Scalar, Self::Dim, Sfx>,
    ) -> Result<(), ProblemError>
    where
        Sx: na::storage::Storage<Self::Scalar, Self::Dim> + IsContiguous,
        Sfx: na::storage::StorageMut<Self::Scalar, Self::Dim>,
    {
        fx[0] = x[0];
        fx[1] = x[1];
        Ok(())
    }
}
```

```
iter = 1	|| fx || = 0.0000000004734010799665374	x = [3.3331393201052606e-10, 3.361702027859792e-10]
solved
```

Whereas this solver has issues:

```rust
struct SolveForOrigin;

impl Problem for SolveForOrigin {
    type Scalar = f32; // CHANGE is here
    type Dim = na::U2;

    fn dim(&self) -> Self::Dim {
        na::U2::name()
    }
}

impl System for SolveForOrigin {
    fn eval<Sx, Sfx>(
        &self,
        x: &na::Vector<Self::Scalar, Self::Dim, Sx>,
        fx: &mut na::Vector<Self::Scalar, Self::Dim, Sfx>,
    ) -> Result<(), ProblemError>
    where
        Sx: na::storage::Storage<Self::Scalar, Self::Dim> + IsContiguous,
        Sfx: na::storage::StorageMut<Self::Scalar, Self::Dim>,
    {
        fx[0] = x[0];
        fx[1] = x[1];
        Ok(())
    }
}
```

```
Error: "neither newton nor steepest descent step can be taken from the point"
```

**Comment by twitchyliquid64** [comment]
I don't think it matters, but my initial parameters were `na::vector![0.23131212, 0.47546]`, and I'm on a Ryzen Threadripper 2920X.

**Comment by pnevyk** [comment]
This is because I used the f64 [epsilon-like constants](https://github.com/datamole-ai/gomez/blob/ee6411ad2c648673a5b3054f5fda8832ddb716aa/src/derivatives.rs#L17-L23) also for f32, which is not great. When I tried to use the correct values for f32, your example solved. I already fixed it in abe0f65, which will be available in the next release. But thank you for showing that it truly matters!

By the way, I see you use `na::U2`. Is it just that you copied it from the provided example or is it important to you to be able to specify a compile-time-known dimension? Because I decided to remove the support for constant dimensions and always use dynamic instead, which simplified the interface a bit and also internals.

**Comment by twitchyliquid64** [comment]
Thanks for the quick fix!!

I'm using dynamic dimensions myself! I just wanted to give you the smallest reproduction case I could :)

**Comment by pnevyk** [comment]
Good :) I will keep the issue open and close it only after I confirm that this example works in the new version once I release it.

**Comment by pnevyk** [comment]
I confirm that this is fixed in the version 0.5.0.
